method to subset and reorder table collection

mufernando opened the pull request:
Here I implemented a numpy version of a method to subset a table collection.

There are probably many uses for this, but most importantly:

- If you subset on a permutation of all existing nodes, you can reorder the Nodes, Individuals, and Populations Tables to your liking. I think @hyanwong was interested in this use?
- For the new `graft` method we need a better way of checking for equivalency between Table Collections, and now we can quickly do so by subsetting.
    - We now can get the non-overlapping bits of two tree sequences using this method, so it could be used to rethink the way `graft` works.

@petrelharp described the motivation and approach [here](https://github.com/tskit-dev/tskit/pull/623#issuecomment-631655756):

The proposed subset-and-reorder method is: tables.subset_nodes(nodes) modifies the tables to retain exactly those:

- nodes that are listed in nodes and in the order listed
- mutations whose node entry is in nodes
- sites with remaining mutations
- edges whose parent and child entries are in nodes
- individuals referred to by the nodes in nodes, and in the order encountered in nodes
- only populations referred to by nodes in nodes, and in the order encountered in nodes
- migrations whose nodes are in nodes

I also implemented a new function to `subset_ragged_col` that subsets ragged cols based on a list of indices (or boolean mask). I added it to `util.py` but not sure that is the right place.

petrelharp commented:
To other viewing this issue - we've got a fair bit of work to tidy this up, so comments appreciated especially on the big picture, but we know that it's not ready to go yet.

Maybe the biggest question in my head is what to call this method? It has two main use cases, really: subsetting, and *reordering*, so maybe the name should reflect reordering also.

petrelharp commented:
To avoid confusion: right now, there are two implementations, one (`.subset()`) in numpy, the other (`.subset_nodes()`) in C. They should be equivalent (except for bugs, of which there are some currently), and it will be interesting to compare them.

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2272,32 +2272,30 @@ class TestSubsetTables(unittest.TestCase):
-        return ts
-
-    def verify_subset_equality(self, tables, subset, nodes):
+    def verify_subset_equality(self, tables, nodes):
```
this is verifying both implementations agree; probably we will stick with the C version, and will put the numpy version in here, to compare against.

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2272,32 +2272,30 @@ class TestSubsetTables(unittest.TestCase):
+        sub2.subset_nodes(nodes)
+        sub1.provenances.clear()
+        sub2.provenances.clear()
+        for x, y in [
```
gee but they don't agree right now, this is debugging code

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2498,18 +2471,35 @@ def drop_index(self):
-        listed, and (vii) the migrations whose nodes are in `nodes`. Note that
-        if `nodes` is a permutation of all nodes in the TableCollection, then
-        the subset operation simply reorders the Nodes, Individuals, and
-        Population Tables in the order encountered in `nodes`.
```
see if you like this documentation?

jeromekelleher commented:
The main big-picture question for me would be, doesn't simplify do this already? We started out calling simplify ``subset``, and that was its first application before we thought about using it for forwards simulation.

hyanwong commented:
> The main big-picture question for me would be, doesn't simplify do this already? We started out calling simplify `subset`, and that was its first application before we thought about using it for forwards simulation.

A good point. Just a quick thought: simplify won't reorder internal nodes arbitrarily (but I'm not sure if the new method will do that either). Also simplify doesn't reorder populations / individuals AFAIK, but I guess it could be extended to do that.

petrelharp commented:
> The main big-picture question for me would be, doesn't simplify do this already?

Simplify does a lot more.  As Yan points out, simplify only lets you reorder samples; but much of the structure above the samples might be changed (nodes removed, edges squashed, etc).  I've put a reminder of this [in the docs](https://github.com/tskit-dev/tskit/pull/663/files#diff-a89cffb041b1b7c33e7be6b595f785a3R2498); should I make this confusing point even more explicit?  If you passed in just the list of sample nodes to subset in an msprime tree, you'd get back the trivial forest where no-one is related to anyone else. Maybe putting a note like this into the docs would help clarify:
```
Note that *only* nodes listed are included: so if full trees are to be retained in the tree sequence, all ancestral node IDs must be included in ``nodes``, not just the samples.
```

jeromekelleher commented:
Ah, that makes sense now. Sure, seems like a useful operation then.

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+        :param list[int] nodes: A list of node IDs to retain in the tables.
+        """
+        if nodes.shape[0] == 0:
+            raise ValueError("Nodes cannot be empty.")
```
Actually, I think it should return an empty tree sequence? (and, the code below should be adjusted to do this if nodes is empty, if it doesn't already).

mufernando replied:
right, we don't need that

jeromekelleher commented:
Ping me when you'd like detailed feedback - I'll mute this thread for now.

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+        if nodes.shape[0] == 0:
+            raise ValueError("Nodes cannot be empty.")
+        if np.max(nodes) >= self.nodes.num_rows:
+            raise ValueError("One of the nodes is not in the TableCollection.")
```
I think the way we usually say this is something like "Node out of bounds". Should also check that all nodes are >= 0?

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+        old_inds = n.individual[nodes]
+        indiv_to_keep, i = np.unique(old_inds, return_index=True)
+        # maintaining order in which they appeard
+        indiv_to_keep = indiv_to_keep[np.argsort(i)]
```
very nice

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+            raise ValueError("Nodes cannot be empty.")
+        if np.max(nodes) >= self.nodes.num_rows:
+            raise ValueError("One of the nodes is not in the TableCollection.")
+        tables = self.copy()
```
I think we should do `self.clear()` at this point.

mufernando replied:
that is not a method of TableCollection. also, that is redundant because we use `set_columns` for all tables.

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+        # removing -1
+        indiv_to_keep = indiv_to_keep[indiv_to_keep != tskit.NULL]
+        # subsetting individuals table
+        if indiv_to_keep.shape[0] == 0:
```
We shouldn't need this special case (esp if we clear the tables above).

mufernando replied:
I think `set_columns` complains about it being empty.

petrelharp commented on python/tskit/util.py:
```diff
@@ -183,6 +183,31 @@ def unpack_arrays(packed, offset):
     return ret
 
 
+def subset_ragged_col(packed, offset, subset, col="metadata"):
```
This is a nice, elegant way of doing what we need. I wonder if should be public, though? Is it too specific to our use case? If it doesn't need to be public, we could call it `_subset_ragged_col`  (and then we wouldn't have to test it). I am honestly not sure. 

I do think the argument `col` should be called `type`, and the method should be called `subset_ragged_column`.

petrelharp commented on python/tskit/util.py:
```diff
@@ -183,6 +183,31 @@ def unpack_arrays(packed, offset):
     return ret
 
 
+def subset_ragged_col(packed, offset, subset, col="metadata"):
```
This is a nice, elegant way of doing what we need! But, should it be public? Or is it too specific to our use case? I honestly don't know. It's a pretty thin wrapper around pack/unpack, but maybe a useful enough one? The return type is a bit odd. Hm - what if it returned a tuple, instead, wouldn't that work? So then we could do e.g. 
```
            self.individuals.set_columns(
                i.flags[indiv_to_keep],
                *util.subset_ragged_col(
                    i.location, i.location_offset, indiv_to_keep, "location"
                ),
                *util.subset_ragged_col(i.metadata, i.metadata_offset, indiv_to_keep),
            )
```
It would mean we couldn't use keywords, which are nice, but the return value would be more pythonic?  If we don't make this public (so, call it `_subset_ragged_col`) then we wouldn't have to test it.

In any case, I think the argument `col` should be called `name`, and the method should be `subset_ragged_column` (ie, with 'column' spelled out).

petrelharp replied:
Oh, and I think we should not provide a default to `col` - there's not a natural default, so the user should specify.

mufernando replied:
changed the function and argument names. the problem with returning a tuple is that `set_columns` only takes keyword arguments.

petrelharp replied:
If you put the arguments in the right order, you don't need to specify the keywords?

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+            )
+        # mapping of ind/pop id in full tables to subset tables
+        ind_map = {ind: i for i, ind in enumerate(indiv_to_keep)}
+        pop_map = {pop: j for j, pop in enumerate(pop_to_keep)}
```
perhaps set `ind_map[tskit.NULL] = tskit.NULL`? (also, you should use NULL, not -1)

petrelharp replied:
and, with that in, I wonder if `np.fromiter(map(ind_map.get, old_inds))` would be more efficient? not quite as clear, though.

petrelharp replied:
Or, wouldn't it be simpler to use the same strategy as `node_map` below, with an array instead of a dict?

mufernando replied:
no, because the individual and population tables could be empty, in which case numpy version would fail

mufernando replied:
also, according to [this](https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array/35216364) not clear if you gain much from using `np.fromiter(map)` compared to a comprehension.

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+        # mapping node ids in full to subsetted table
+        # making the node map +1 bc last will be mapping -1 to -1
+        node_map = np.repeat(tskit.NULL, tables.nodes.num_rows + 1)
+        node_map[-1] = tskit.NULL
```
this line is unnecessary

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+            site_map = site_map.astype("int32")
+            mutation_map = np.repeat(tskit.NULL, tables.mutations.num_rows)
+            mutation_map[keep_muts] = np.arange(np.count_nonzero(keep_muts))
+            mutation_map = mutation_map.astype("int32")
```
you should be able to put the dtype in the initial line (np.repeat)?

petrelharp commented on python/tskit/tables.py:
```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
+            # adding tskit.NULL to the end to map -1 -> -1
+            mutation_map = np.concatenate(
+                (mutation_map, np.array([tskit.NULL], dtype="int32"))
+            )
```
do like you did above - just initialize the map to be one longer

mufernando commented:
> I don't think the `subset_ragged_columns` is working as it should because it is not reordering.

fixed this and added metadata to the tests

mufernando commented:
Speed comparisons for the C and Numpy versions:

Here for one tree of 4M nodes, I tested the speed for different sizes of subset:
<img width="791" alt="Screen Shot 2020-06-03 at 2 23 52 PM" src="https://user-images.githubusercontent.com/9950100/83668118-db8f2600-a5a5-11ea-8c2d-514edcd1b961.png">

Here for trees of varying sizes, I tested the speed to get a same sized subset (40 nodes):
<img width="802" alt="Screen Shot 2020-06-03 at 2 24 27 PM" src="https://user-images.githubusercontent.com/9950100/83668177-efd32300-a5a5-11ea-9f2b-6921db97fd78.png">

mufernando commented:
add a test with random locations

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2272,19 +2272,231 @@ class TestSubsetTables(unittest.TestCase):
+        return pack(unpacked)
+
+    def np_subset(self, tables, nodes, record_provenance=True):
+        """
```
We don't want to duplicate documentation. Just leave this at "Numpy version of TableCollection.subset, for testing purposes."

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2272,19 +2272,231 @@ class TestSubsetTables(unittest.TestCase):
+            raise ValueError("Nodes cannot be empty.")
+        if np.max(nodes) >= tables.nodes.num_rows or np.any(nodes < 0):
+            print(nodes)
+            print(np.all(nodes > 0))
```
remove print statements

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+        ind_map = np.repeat(tskit.NULL, tables.individuals.num_rows + 1)
+        ind_map[indivs] = np.arange(len(indivs), dtype="int32")
+        pop_map = np.repeat(tskit.NULL, tables.populations.num_rows + 1)
+        pop_map[pops] = np.arange(len(pops), dtype="int32")
```
How about instead:
```
        ind_map = np.repeat(tskit.NULL, tables.individuals.num_rows + 1)
        pop_map = np.repeat(tskit.NULL, tables.populations.num_rows + 1)
        num_inds = 0
        num_pops = 0
        for k, n in enumerate(nodes):
            node_map[n] = k
            ind = tables.nodes[n].individual
            pop = tables.nodes[n].population
            if ind not in indivs and ind != tskit.NULL:
                ind_map[ind] = num_inds
                num_inds += 1
            if pop not in pops and pop != tskit.NULL:
                pop_map[pop] = num_pops
                num_pops += 1
```

petrelharp replied:
oh wait never mind, you use pops and indivs below! =)

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+        tables2 = tables.copy()
+        tables2.subset_nodes(np.arange(ts.num_nodes))
+        # except that you need one extra row in the provenances
+        self.assertEqual(tables.provenances.num_rows + 1, tables2.provenances.num_rows)
```
how about do record_prov=False and not need to clear the provenances?

mufernando replied:
it seems the default in other tskit functions is `record_provenance=True` so I wanted to keep it consistent.

petrelharp replied:
That should dfeinately be the default! I was saying that in this call to subset, you ask it to not record provenance.

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+        tables = ts.tables
+        tables2 = tables.copy()
+        tables2.subset_nodes(np.arange(ts.num_nodes - 1, -1, -1))
+        tables.provenances.clear()
```
same here?

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+                Ne=100, sample_size=10, mutations_per_branch=num_muts
+            )
+            ts = tsutil.add_random_metadata(ts, self.seed)
+            # adding locations
```
don't need this comment, really

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+            ts = self.get_msprime_mig_example(
+                Ne=100, sample_size=10, mutations_per_branch=num_muts
+            )
+            ts = tsutil.add_random_metadata(ts, self.seed)
```
how about put the random metadata and individuals up in get_mig_example?

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,374 @@ def test_set_columns_not_implemented(self):
+
+        **Note:** This is quite different from :meth:`.simplify`: the resulting
+        tables contain only the nodes given, not ancestral ones as well, and
+        does not simplify the relationships in any way.
```
How about adding:
```
To subset the tables to a portion of the genome, see :meth:`.keep_intervals`.
```

petrelharp commented:
Looks good to me!  Questions for others:

- How's the `subset( )` name? We might want to subset by other things in the future, but that could be done by adding an argument `individuals=`, for instance.

- Should the numpy version remain in the testing code? It's nice to have the independent implementation, but it's a kinda bulky one, so maybe it doesn't need to be here? It seems OK to me, if slightly more than is necessary.

hyanwong commented:
> * How's the `subset( )` name? We might want to subset by other things in the future, but that could be done by adding an argument `individuals=`, for instance.

What alternatives are there? E.g. is `reduce` any better? I think it's hard to decide until we have a list of candidates.

> * Should the numpy version remain in the testing code? It's nice to have the independent implementation, but it's a kinda bulky one, so maybe it doesn't need to be here? It seems OK to me, if slightly more than is necessary.

I don't think we care too much about bulking up the testing code, do we? Better to have it *somewhere*, for future ref (it might not be that obvious to dig it up from the history)

petrelharp commented:
> What alternatives are there? E.g. is reduce any better? I think it's hard to decide until we have a list of candidates.

I'm looking for some. But if "subset" works, no need to look hard! I like subset better than reduce. slice_nodes occurs to me, but doesn't do it for me either. 'reorder' maybe, but then that doesn't connote the subsetting operation.

mufernando commented:
I think we are ready for your more detailed comments, @jeromekelleher!

jeromekelleher commented:
> I think we are ready for your more detailed comments, @jeromekelleher!

Thanks @mufernando! Sorry I haven't had a chance to look at this yet, and am away tomorrow, so it'll probably be next week before I have a look.

petrelharp commented:
A thought here: we should probably note in the docs that if the nodes are divided into disjoint subsets A and B, then tables.subset(A) and tables.subset(B) don't quite contain everything in the original tables, because they'd miss out edges between A and B.

jeromekelleher commented:
Sorry, I being so slow to look here - I promise I'll do it tomorrow. In the meantime, here's an example that came up from an msprime user who wanted to simulate populations with different recombination maps:
```python
L = 1000
T = 1
map0 = msprime.RecombinationMap.uniform_map(rate=0.01, length=L, discrete=True)
map1 = msprime.RecombinationMap.uniform_map(rate=0.02, length=L, discrete=True)
population_configurations = [
    msprime.PopulationConfiguration(),
    msprime.PopulationConfiguration()
]

# Simulate population 0
ts_pop0 = msprime.simulate(
    samples=[msprime.Sample(population=0, time=0)] * 10,
    population_configurations=population_configurations,
    recombination_map=map0, end_time=T)
# Simulate population 1
ts_pop1 = msprime.simulate(
    samples=[msprime.Sample(population=1, time=0)] * 10,
    population_configurations=population_configurations,
    recombination_map=map1, end_time=T)

t0 = ts_pop0.tables
t1 = ts_pop1.tables

# Insert the nodes for t1 into t0
N = len(t0.nodes)
t0.nodes.append_columns(
    flags=t1.nodes.flags,
    time=t1.nodes.time,
    population=t1.nodes.population,
    individual=t1.nodes.individual,
    metadata=t1.nodes.metadata,
    metadata_offset=t1.nodes.metadata_offset)

# Add the edges for t1 into t0
t0.edges.append_columns(
    left=t1.edges.left,
    right=t1.edges.right,
    parent=t1.edges.parent + N,
    child=t1.edges.child + N)

t0.sort()
joint_ts = t0.tree_sequence()

# Now run the simulation of the joint history.
final_ts = msprime.simulate(
    from_ts=joint_ts,
    population_configurations=population_configurations,
    # Migrate everyone to population 0
    demographic_events=[
        msprime.MassMigration(time=T, source=1, dest=0)
    ],
    recombination_map=map0)
```

The idea is we simulate the two populations independently, then merge the two tree sequences together and then simulate their shared ancestry. This is quite similar to what's going on here, right? Should we also provide an operation to ``add`` two tree sequences together?

petrelharp commented:
> This is quite similar to what's going on here, right? Should we also provide an operation to add two tree sequences together?

I think you meant to post this example over in #623?  In which we have decided (I think) to call the method `union()`, not `graft()`?

jeromekelleher commented:
> I think you meant to post this example over in #623? In which we have decided (I think) to call the method union(), not graft()?

You're right, I did. Crud. I've moved the example over there for reference.

jeromekelleher commented:
I will definitely comment on this PR first thing tomorrow!

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    int ret = 0;
+    tsk_id_t pop, ind;
+    tsk_size_t metadata_length;
+    if (row >= (tsk_id_t) source->num_rows) {
```
Need to check if ``row < 0`` as well

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        goto out;
+    }
+    pop = source->population[row];
+    if (pop >= 0) {
```
pop != TSK_NULL is clearer, I think.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+{
+    int ret = 0;
+    tsk_id_t pop, ind;
+    tsk_size_t metadata_length;
```
Actually, it'd save a bit of trouble if you did
```c
ret = node_table_get_row(source, row, &node);
if (ret < 0) {
    goto out;
}
```
This will do the bounds checking for you, and save some tedium in dealing with the offsets.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
+    tsk_table_collection_t tables;
+
+    node_map = calloc(self->nodes.num_rows, sizeof(tsk_id_t));
```
Since we're memsetting all of these straight afterwards, might as well use ``malloc`` here.

jeromekelleher replied:
Also, better do use sizeof(*node_map), e.g.
```c
node_map = malloc(self->nodes.num_rows * sizeof(*node_map));
```
this is a bit more robust and less error prone.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+{
+    int ret = 0;
+    tsk_id_t k, n, m, ind, pop, mut, site;
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
```
So it's not essential here as we're mallocing these pointers before we do anything else, but I think it's still better to explicitly inititialise these variables to NULL. We might update the code later and put in (say) an error check before the mallocs and then everything would explode. So,
```
tsk_id_t *node_map = NULL;
tsk_id_t *individual_map = NULL; // etc
```

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+
+int TSK_WARN_UNUSED
+tsk_table_collection_subset(
+    tsk_table_collection_t *self, tsk_id_t *nodes, size_t num_nodes)
```
I'd say ``tsk_size_t`` is more appropriate for ``num_nodes``.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        n = nodes[k];
+        node_map[n] = k;
+        ind = tables.nodes.individual[n];
+        if (ind >= 0) {
```
I think ind != TSK_NULL is a bit clearer. Ditto below.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        node_map[n] = k;
+        ind = tables.nodes.individual[n];
+        if (ind >= 0) {
+            if (individual_map[ind] < 0) {
```
``if individual_map[ind] == TSK_NULL`` - more concrete, clearer to the reader.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+
+    // edges
+    for (k = 0; k < (tsk_id_t) tables.edges.num_rows; k++) {
+        m = node_map[tables.edges.parent[k]];
```
I'm not sure we're getting much from the ``copy_row`` function here. What about
```c
edge_table_get_row(&tables.edges, k, &edge);
new_parent = node_map[edge.parent];
new_child = node_map[edge.child];
if (new_parent != TSK_NULL and new_child != TSK_NULL) {
    ret = edge_table_add_row(&self->edges,
          edge->left, edge->right, new_parent, new_child, edge->metadata, edge->metadata_length);
    if (ret < 0) {
        goto out;
    }
```

This actually makes life quite a bit simpler, because we don't need to test the ``edge_table_copy_row`` function now.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        n = nodes[k];
```
I find the notation ``n`` for a node confusing. We use variables like ``u`` and ``v`` to refer to nodes elsewhere, and keep ``n`` for things like "this is a fixed number".

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+
+    // mutations and sites
+    mut = 0;
+    for (site = 0; site < (tsk_id_t) tables.sites.num_rows; site++) {
```
I think it would be simpler to use the ``get_row`` functions here and skip the helpers, like
```c
for (site_id = 0; ...) {
     tsk_site_table_get_row(&tables.sites, site_id, &site);
     while ((mutation_id < (tsk_id_t) tables.mutations.num_rows)
             && (tables.mutations.site[mutation_id] == site)) {
             tsk_mutation_table_get_row(&tables.mutations, mutation_id, &mutation)
             new_node = node_map[mutation.node];
             if (new_node != TSK_NULL) {
                   if (site_map[node.site] == TSK_NULL) {
                          ret = tsk_site_table_add_row(&tables.sites, site.position, /* etc */);                
                          if (ret < 0) {
                                goto out;
                          }
                         site_map[site] = ret;
                    }
                    // etc
             }
}

jeromekelleher commented on c/tskit/tables.h:
```diff
@@ -2409,6 +2409,24 @@ completes.
+@brief Subsets a table collection by a set of nodes.
+
+@rst
+Reduces the table collection, retaining only the nodes given in ``nodes``,
```
This is a bit terser than the Python docs. C programmers will want precision too. I think we want something a more like the version we have on the python TreeSequence documentation.

jeromekelleher commented on python/_tskitmodule.c:
```diff
@@ -6256,6 +6256,38 @@ TableCollection_link_ancestors(TableCollection *self, PyObject *args, PyObject *
+        goto out;
+    }
+    ret = Py_BuildValue("");
+out:
```
Very nice, this is perfect. We need some tests of this in test_lowlevel.py though, to make sure that the error conditions are handled correctly. We'll want to call the new method with, e.g., the incorrect number of arguments and bad types for numpy to make sure these error conditions are handled.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    }
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
```
So if the first thing we do here is 
```
      ret = tsk_node_table_get_row(&tables.nodes, nodes[k], &node);
      if (ret < 0) {
           goto out;
      }
```
we do two things: 1) we check input, to make sure it's a valid node reference; and (2) we get the actual node object.

jeromekelleher replied:
We'll definitely want some tests to make check that we error out when we pass in bad node refs.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
```
The behaviour is currently undefined if any of these mallocs fail. That's because the ``tables`` variable is uninitialised, and we call table_collection_free on it below. We need to either memset the tables to zero, or run the table collection copy *first* before the mallocs (assuming they've been set to NULL).

mufernando replied:
ok, so memset tables to zero right after declaration

jeromekelleher replied:
Yeah, either that or run tsk_table_collection_copy before the mallocs. I think this is guaranteed to init the object.

mufernando commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2266,336 @@ def test_set_columns_not_implemented(self):
+        unpacked = [unpacked[i] for i in subset]
+        return pack(unpacked)
+
+    def np_subset(self, tables, nodes, record_provenance=True):
```
I agree this naive version would be the best as a test, but note we have the `verify_subset` which is more obviously correct. There we go row by row for each table.

I don't think we need the `np_subset` in the test to make it robust, but we decided to keep it there in case we want to have a look at it in the future. Importantly, with the numpy version we discovered that even though the numpy version is a little easier to write, the C version is waay faster.

petrelharp replied:
> Importantly, with the numpy version we discovered that even though the numpy version is a little easier to write, the C version is waay faster.

This was a good exercise (for me at least) - naively, I thought that the numpy version would be easier to write and just as fast; in fact, it was harder to write and slower. (see the graph above) Maybe we could make it just as fast with additional work, but that would defeat the purpose. I lean towards leaving it in but would support removing it if you have a stronger opinion.

jeromekelleher replied:
Which graph is this @petrelharp? 

Sure, we can keep the numpy version in, but I'd still like to see the naive dumb-as-rocks Python row-by-row version. I agree that the verify function does this, but it's not quite the same thing to check something as it is to write it down and actually do it. We can always compare all three versions to each other. In terms of making sure that the ragged column stuff is done properly, I think this is really the best and simplest way to do it.

petrelharp replied:
Took me a minute to find, but [here](https://github.com/tskit-dev/tskit/pull/663#issuecomment-638340711) it is.

I'm not yet convinced that the naive version would make the testing more robust here, but I'll discuss with @mufernando.

jeromekelleher replied:
> Took me a minute to find, but here it is.

Thanks - that is an impressive difference. Writing non-simple things in numpy is hard!

> I'm not yet convinced that the naive version would make the testing more robust here, but I'll discuss with @mufernando.

It's not just about testing robustness @petrelharp, it's about clarity of what the test is actually *doing*. We need to be able to read these tests in 5 years time and understand how they work - hence, my push for a simple, direct, Python implementation.

We'd have written it in the time it takes to write these comments...

mufernando commented:
I think I addressed all of your comments, @jeromekelleher 

I was fighting with codecov bc it says the loops over mutations and migrations are never covered by our tests, but I'm pretty sure they are -- all our examples have mutations and migrations which seem to be correctly subsetted...

petrelharp commented:
I can take a look first if you like, @jeromekelleher.

jeromekelleher commented:
> I can take a look first if you like, @jeromekelleher.

Sure, that'd be great, thanks @petrelharp.  Or are you procrastinating from something...

petrelharp commented on c/tskit/tables.h:
```diff
@@ -2409,6 +2409,36 @@ completes.
+3. Edges: if both parent and child are retained nodes.
+4. Mutations: if the mutation's node is a retained node.
+5. Sites: if any mutations remain at the site after removing mutations.
+6. Migrations: if the migration's node is a retained node.
```
Mabye add afterwards: "Retained edges, mutations, sites, and migrations appear in the same order as in the original tables."

petrelharp commented on python/tests/test_lowlevel.py:
```diff
@@ -233,9 +233,9 @@ def test_simplify_bad_args(self):
         tc = ts.tables.ll_tables
         with self.assertRaises(TypeError):
-            tc.simplify()
+            tc.subset()
```
whoops! did you mean to copy this test?

petrelharp replied:
Oh, I see you did, below.

mufernando replied:
oops, fixed it

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2268,169 @@ def test_set_columns_not_implemented(self):
+
+    def test_wf_examples(self):
+        tables = wf.wf_sim(5, 5, seed=65634)
+        tables.sort()
```
why the sort?

mufernando replied:
I'm not sure, but all the other examples with the WF simulator do that. I think it has to do with the Edges Tables requirements.

petrelharp replied:
Ah, well, since we're just working with tables, not tree sequences, let's *not* sort!

mufernando replied:
We need tree sequences for some of the tsutil functions (to add metadata, etc).

petrelharp replied:
Hm: those don't need the tree sequence either, but ok!

mufernando replied:
what do you mean? they take in tree sequences and dump the tables internally.

petrelharp replied:
Right - they could be rewritten to operate on tables instead of tree sequences, since that's really what's going on. But, I'm not suggesting you do that.

petrelharp commented on python/tskit/trees.py:
```diff
@@ -4401,6 +4401,41 @@ def trim(self, record_provenance=True):
+        3. Edges: if both parent and child are retained nodes.
+        4. Mutations: if the mutation's node is a retained node.
+        5. Sites: if any mutations remain at the site after removing mutations.
+        6. Migrations: if the migration's node is a retained node.
```
same note as in the C docs

petrelharp commented on python/tests/test_tables.py:
```diff
@@ -2266,3 +2268,169 @@ def test_set_columns_not_implemented(self):
+            nodes = rng.choice(np.arange(ts.num_nodes), n, replace=False)
+            self.verify_subset(tables, nodes)
+
+    def test_wf_examples(self):
```
How about move the WF code to `self.get_wf_example`, and then in all the tests you can do something like this:
```
def get_examples():
   yield self.get_msprime_mig_example(Ne=100, sample_size=10, seed=12396)
   yield self.get_wf_example()

def test_subset_all(self):
    for ts in self.get_examples():
```

That way we'd be running the tests on more than one example, and would separate out the example generation from the tests.

mufernando replied:
good idea!

petrelharp commented:
Looks good - just a few comments. I don't understand codecov either.

mufernando commented:
I think I addressed all the issues. Not sure about codecov, @jeromekelleher. I don't understand how it works -- sometimes it complains sometimes it doesn't. Pretty sure the parts about mutations and migrations in the C code are covered by our tests, though.

petrelharp commented:
Yep, over to you, @jeromekelleher (except my note about removing the sort() above).

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        if (ret < 0) {
+            goto out;
+        }
+        iid = tables.nodes.individual[nid];
```
No need to look up the tables here - it's  more efficient to use the ``node_t`` struct that you've created. So, here,
```c
if (node.individual != TSK_NULL) {
     if (individual_map[node.individual] == TSK_NULL) {
```
which is also more readable.

jeromekelleher replied:
Ah, I see, you were also handling the case there the individual is NULL. How about,
```c
new_ind_id = TSK_NULL;
if (node.individual != TSK_NULL) {
     // as before, then
     new_ind_id = individual_map[node.individual];
}

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+                }
+                individual_map[iid] = ret;
+            }
+            iid = individual_map[iid];
```
This isn't very readable - iid is referring to two different IDs at different points in the loop. How about ``new_ind_id`` or something?

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+            }
+            iid = individual_map[iid];
+        }
+        pid = tables.nodes.population[nid];
```
Can use the same pattern here,
```c
new_pop_id = TSK_NULL;
if (node.population != TSK_NULL) {
    // etc
    new_pop_id = population_map[node.population];
}

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        if (ret < 0) {
+            goto out;
+        }
+        if (ret != k) {
```
Why would ret not be equal to k? If this is checking for the presence of a bug it should be an assert (but, I think it's unnecessary?, could just assert this at the end of the loop once?

petrelharp replied:
> could just assert this at the end of the loop once?

good call, yes

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        nid = nodes[k];
```
I wouldn't bother saving this variable, just run  ``tsk_node_table_get_row(&tables.nodes, nodes[k], &node);`` and use ``node.id`` everywhere later. This is exactly as efficient (possibly more, as you're reducing the number of variables in the stack, and therefore better for cache).

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    // edges
+    for (eid = 0; eid < (tsk_id_t) tables.edges.num_rows; eid++) {
+        tsk_edge_table_get_row(&tables.edges, eid, &edge);
+        pid = node_map[tables.edges.parent[eid]];
```
We don't need to look up the table here, how about 
```c
new_parent = node_map[edge.parent];
new_child = node_map[edge.parent];
```
It's more readable than pid and eid.  (What's p, and e? Which set of tables is it an ID in?)

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        while ((mid < (tsk_id_t) tables.mutations.num_rows)
+               && (tables.mutations.site[mid] == sid)) {
+            tsk_mutation_table_get_row(&tables.mutations, mid, &mut);
+            nid = node_map[mut.node];
```
new_node, etc.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    }
+
+    // mutations and sites
+    mid = 0;
```
``mutation_id`` is better - is this the ``mid`` point of a line?

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        if (nid != TSK_NULL) {
+            sid = mig.source;
+            did = mig.dest;
+            if (sid != TSK_NULL) {
```
Migrations are guaranteed to be non-null, so you can skip this check.

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+            if (sid != TSK_NULL) {
+                sid = population_map[sid];
+            }
+            if (did != TSK_NULL) {
```
Actually, what if this is a population we haven't already seen before? You would need to check the map, I guess, and update if necessary. I don't think I'd bother though: unless you have an application using migrations, I would just raise an error if there's stuff in the migration table. I don't think it's really used much. This is what we do in simplify fwiw (there's an error code for "we don't support migrations" which could be reused).

mufernando replied:
Right, the dest population has to be in the map, but it is possible the source isn't. I didn't know source and dest had to be non-NULL. My idea was that in this case the source pop would be NULL in the subset tables. Alternatively we could remove migrations with source pop not seen in any of the `nodes` from the subset. I think we are so close to being able to support the Migrations Table in this operation that we should just go for it.

jeromekelleher replied:
I think either of them could be missing from the map. Suppose you have a node that starts in one population then goes for a long journey through hundreds of populations. In that case the majority of the populations will not have been seen before. So, I think we'd have to append all these populations to the map.

There's a fair bit of testing to be done in this... It would be a lot quicker the other way.

petrelharp replied:
Ah, of course - I'd intended them to be NULL in this case, but of course it doesn't make sense for them to be NULL. I agree that throwing an error is the thing to do.

petrelharp replied:
Whoops - i meant to throw an error if the source or dest turns out to be unmapped.

mufernando replied:
I decided to remove support for migrations altogether...

jeromekelleher commented on python/tests/tsutil.py:
```diff
@@ -539,6 +539,85 @@ def compute_mutation_parent(ts):
+                mig.left,
+                mig.right,
+                node_map[mig.node],
+                pop_map.get(mig.source, tskit.NULL),
```
I don't think this works - see comments above in the C. I think we should just not bother with migrations.

jeromekelleher commented:
We also need to bump up test coverage quite a bit. The C tests don't have any sites or mutations, so that's not being covered. Neither are migrations (but we should just drop that anyway, and test that we behave correctly when we try to subset a including them).

Need to add a test for the TreeSequence.subset method also. This can just be verifying that it returns the same thing as ``ts.dump_tables().subset()``

mufernando commented:
thanks for the revisions, @jeromekelleher! I added some mutations to the C side, but codecov isn't being rerun. Can we explicitly ask it to be run again?

benjeffery commented:
Codecov isn't reporting as valgrind is reporting a memory leak: https://circleci.com/gh/tskit-dev/tskit/2088?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

mufernando commented:
yeah, just fixed that. but either way codecov is not in the list of pending checks.

benjeffery commented:
Codecov will only appear after circle ci is successfully completed, the coverage is measured in the circle CI run.

jeromekelleher commented:
Looking great, thanks @mufernando! Apologies for all the nitpicking, but I think it was worth the effort. The code is super-clear and clean now.

Just need to get codecov happy and squash the commits now I think.

mufernando commented:
Thank you, @jeromekelleher and @petrelharp for the patience and help along the way!

I am not sure how to bump up codecov at this point. The parts of the code not being tested are some ifs on the return of `add_row` lines. What is the proper way of dealing with this?

jeromekelleher commented:
You've tested all the conditions that can reasonably happen @mufernando, so we're ready to merge. Can you squash the commits please?

mufernando commented:
woot woot!

thanks, you all!
